### PR TITLE
Migrate forward plugin to v0.14 api

### PR DESCRIPTION
**Pull request by tagomoris**
This change is to migrate in/out_forward to v0.14 APIs, except for socket/server plugin helpers, to keep commits few and diff small.

### CODE REVIEW

**Comment by tagomoris** [comment]
Pull requests to add socket/server plugin helpers will follow this change.
@repeatedly Could you review this?

**Comment by repeatedly** [comment]
Okay. I will review it later 👍

**Comment by tagomoris** [comment]
@repeatedly ping?

File: lib/fluent/plugin/in_forward.rb [hunk 1]
```diff
@@ -125,46 +128,30 @@ def configure(conf)
...
-      # This problem occurs possibly because those watchers are thought to be related to each other
-      # and when detaching one of them the other is also detached for some reasons.
-      # As a workaround, check if watchers are attached before detaching them.
-      @loop.watchers.each {|w| w.detach if w.attached? }
-      @loop.stop
-      @usock.close
-      @thread.join
-      @lsock.close
-
+    def stop
```
Comment (repeatedly) [inline lib/fluent/plugin/in_forward.rb]: `stop` is called before `shutdown`, right?
What happens when close socket before shutdown/close event loop?
Is it safe?

File: lib/fluent/plugin/in_forward.rb [hunk 2]
```diff
@@ -14,23 +14,26 @@
...
-require 'cool.io'
+require 'fluent/plugin/input'
+require 'fluent/msgpack_factory'
 require 'yajl'
+require 'digest'
 
-require 'fluent/input'
+require 'fluent/plugin/socket_util'
+require 'fcntl'
+require 'cool.io'
```
Comment (repeatedly) [inline lib/fluent/plugin/in_forward.rb]: `cool.io` is required by event_loop helper.

Comment (repeatedly) [inline lib/fluent/plugin/in_forward.rb]: Ah wait. This commit doesn't include socket helpers. Ignore this comment.

File: lib/fluent/plugin/out_forward.rb [hunk 3]
```diff
@@ -14,42 +14,23 @@
...
+require 'fluent/compat/socket_util'
 
-  class ForwardOutputACKTimeoutError < ForwardOutputResponseError
-  end
+module Fluent::Plugin
+  class ForwardOutput < Output
+    class Error < StandardError; end
+    class ResponseError < Error; end
+    class ConnectionClosedError < Error; end
+    class ACKTimeoutError < Error; end
```
Comment (repeatedly) [inline lib/fluent/plugin/out_forward.rb]: In v0.12, ForwardOutputACKTimeoutError is subclass of ForwardOutputResponseError.
Change from ResponseError to Error is intented?

File: test/plugin/test_in_forward.rb [hunk 4]
```diff
@@ -74,148 +88,153 @@ def test_auth
...
+      begin
+        Timecop.freeze(Time.at(time))
+        @d = d = create_driver
+
+        records = [
+          ["tag1", time, {"a"=>1}],
+          ["tag2", time, {"a"=>2}],
+        ]
+
+        d.run(expect_records: records.length, timeout: 5) do
```
Comment (repeatedly) [inline test/plugin/test_in_forward.rb]: Why change timeout to 5?

File: lib/fluent/plugin/out_forward.rb (hunk 3 above)
Comment (tagomoris) [inline lib/fluent/plugin/out_forward.rb]: It's intended. There are no actual benefits to make these errors with sub-super classes.
And, in fact, there are no responses under ACK timeout situation.d

File: test/plugin/test_in_forward.rb [hunk 5]
```diff
@@ -74,148 +88,153 @@ def test_auth
...
         ["tag1", time, {"a" => 1}],
         ["tag2", time, {"a" => 2}],
       ]
 
-      d.run do
-        entries = records.map { |tag, _time, record| [tag, _time, record] }
+      d.run(shutdown: false, expect_records: 2, timeout: 10) do
+        entries = []
         # These entries are skipped
         entries << ['tag1', true, {'a' => 3}] << ['tag2', time, 'invalid record']
```
Comment (repeatedly) [inline test/plugin/test_in_forward.rb]: Why use 2 lines intead of `entries = [[...], [...]]`?

File: test/plugin/test_in_forward.rb (hunk 4 above)
Comment (tagomoris) [inline test/plugin/test_in_forward.rb]: Just for stability of CI tests.

File: test/plugin/test_in_forward.rb (hunk 5 above)
Comment (tagomoris) [inline test/plugin/test_in_forward.rb]: To reduce the number of lines of diff.

File: test/plugin/test_out_forward.rb [hunk 6]
```diff
@@ -29,9 +32,13 @@ def setup
   ]
 
   def create_driver(conf=CONFIG)
-    d = Fluent::Test::BufferedOutputTestDriver.new(Fluent::ForwardOutput) {
+    Fluent::Test::Driver::Output.new(Fluent::Plugin::ForwardOutput) {
       attr_reader :responses, :exceptions
 
+      def write(chunk)
```
Comment (repeatedly) [inline test/plugin/test_out_forward.rb]: Why is this definition needed?

File: lib/fluent/plugin/out_forward.rb (hunk 3 above)
Comment (repeatedly) [inline lib/fluent/plugin/out_forward.rb]: okay

**Comment by repeatedly** [comment]
Commented. Others looks good.

File: test/plugin/test_out_forward.rb (hunk 6 above)
Comment (tagomoris) [inline test/plugin/test_out_forward.rb]: `Output#implement?` checks the features of that plugin by checking instance methods implemented in just that class, without methods of superclasses.
This test code defines a one-time class inherits ForwardOutput. If this method is not defined, `Output#implement?` returns false for check of buffered output feature.

It's better to be fixed in future, but currently, this code works fine.

Comment (repeatedly) [inline test/plugin/test_out_forward.rb]: I see. I understood the situation.

**Comment by tagomoris** [comment]
I pushed a commit to close sockets in `#close`, and will merge this after CI green.

**Pull request merged**